### PR TITLE
PullRequest for Issue103: Using signal map for AMDS scaler to identify the step scan finished

### PR DESCRIPTION
**Pull request by XuetaoLiu-S**
@davidChevrier @dretrex @iainworkman 

Hi guys, pls take a look at this PR. This supposed to fix the issues. I'm running the test right now.

There are also code refactors required for this. There are some redundant codes there to be optimized. I will do that in a sub branch.

### CODE REVIEW

**Comment by XuetaoLiu-S** [comment]
No rush... the step scan failed...

**Comment by XuetaoLiu-S** [comment]
Scan is running now with restarting Acquaman

**Comment by dretrex** [comment]
Once the scans are working a bit better, post in the thread again.  Then we can go over the completed code.

**Comment by dretrex** [comment]
Does that mean we are good for review?

File: source/Detector/Scaler/AMDSScalerDetector.cpp [hunk 1]
```diff
@@ -265,11 +269,79 @@ void AMDSScalerDetector::onTriggerDwellTimerTimeout()
...
 
+void AMDSScalerDetector::startScalerDwellAcquisition()
+{
+	double seconds = triggerDwellInterfaceDwellTimeControl_->value();
+	if(seconds > 0 && seconds < 100) {
+		if (triggerDwellInterfaceDwellModeControl_->withinTolerance(0.0)) {
+			// set the mapping of value changed. NOTE: This should be in a more graceful way
+			waitingChannels_.clear();
+
+			QMap<quint8, int> channelToIndexMap;
```
Comment (XuetaoLiu-S) [inline source/Detector/Scaler/AMDSScalerDetector.cpp]: This section of codes are stolen from the `setFlattenedData()`. 

It will be optimized later

**Comment by XuetaoLiu-S** [comment]
@dretrex yes, u can review now

File: source/Detector/Scaler/AMDSScalerDetector.cpp [hunk 2]
```diff
@@ -229,25 +233,25 @@ void AMDSScalerDetector::onFetchScanBuffer()
 	}
 }
 
+void AMDSScalerDetector::onTriggerDwellInterfaceDwellModeControlValueChanged(double value)
+{
+	Q_UNUSED(value)
+
+	if(triggerDwellInterfaceDwellModeControl_->withinTolerance(1.0) && triggerDwellInterfaceStartControl_->withinTolerance(1.0)) {
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetector.cpp]: Just wanting to double check that these are both supposed to be at 1 if we are wanting to stop acquisition?

File: source/Detector/Scaler/AMDSScalerDetector.cpp [hunk 3]
```diff
@@ -229,25 +233,25 @@ void AMDSScalerDetector::onFetchScanBuffer()
...
-			double seconds = triggerDwellInterfaceDwellTimeControl_->value();
-			if(seconds > 0 && seconds < 100){
-				int asMsecs = (int)(seconds*1000);
-				triggerDwellTimer_->setInterval(asMsecs);
-				triggerDwellInterfaceDwellStateControl_->move(1.0);
-				triggerDwellTimer_->start();
-			}
+			startScalerDwellAcquisition();
 		}
 		else if(triggerDwellInterfaceStartControl_->withinTolerance(0.0) && triggerDwellInterfaceDwellModeControl_->withinTolerance(1.0) && triggerDwellTimer_->isActive()){
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetector.cpp]: Is this supposed to be different from the line where my earlier comment is?

File: source/Detector/Scaler/AMDSScalerDetector.cpp [hunk 4]
```diff
@@ -265,11 +269,79 @@ void AMDSScalerDetector::onTriggerDwellTimerTimeout()
...
 		triggerDwellTimer_->stop();
-		triggerDwellInterfaceStartControl_->move(0);
 	}
 	emit requestFlattenedData(triggerDwellInterfaceDwellTimeControl_->value());
 }
 
+void AMDSScalerDetector::startScalerDwellAcquisition()
+{
+	double seconds = triggerDwellInterfaceDwellTimeControl_->value();
+	if(seconds > 0 && seconds < 100) {
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetector.cpp]: Is there a reason for the `< 100` part of the conditional?

File: source/Detector/Scaler/AMDSScalerDetector.cpp [hunk 5]
```diff
@@ -265,11 +269,79 @@ void AMDSScalerDetector::onTriggerDwellTimerTimeout()
...
+		disconnect(channelValueChangedSignalMapper_, SIGNAL(mapped(int)), this, SLOT(onScalerChannelValueChanged(int)));
+
+		triggerDwellInterfaceStartControl_->move(0);
+		triggerDwellInterfaceDwellStateControl_->move(0);
+	}
+}
+
+void AMDSScalerDetector::onScalerChannelValueChanged(int channelIndex)
+{
+	if(waitingChannels_.contains(channelIndex)){
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetector.cpp]: Does this function only remove channels?  If so, I think this function should be renamed.

**Comment by dretrex** [comment]
Just a couple comments.
Time Estimate: 30 minutes

**Comment by dretrex** [comment]
Overall, this still looks good.  Only thing now is to really make sure that we put it through its paces when we test it.

**Comment by dretrex** [comment]
@XuetaoLiu-CLS, those were the wrong const uses.  They are const functions: `bool isAcquiring() const;` not `const bool isAcquiring();`

**Pull request merged**